src/dom/Decimal.ts

```diff
@@ -56,4 +57,44 @@ export class Decimal extends Value(
     writer.setAnnotations(this.getAnnotations());
     writer.writeDecimal(this.decimalValue());
   }
+
+  _ionEquals(
+    expectedValue: any,
+    options: {
+      epsilon?: number | null;
+      ignoreAnnotations?: boolean;
+      ignoreTimestampPrecision?: boolean;
+      onlyCompareIon?: boolean;
+    } = {
+      epsilon: null,
+      ignoreAnnotations: false,
+      ignoreTimestampPrecision: false,
+      onlyCompareIon: true,
+    }
+  ): boolean {
+    // check if it is a valid JS Object or dom.Value Object based on options.onlyCompareIon value
+    if (
+      !(
+        (options.onlyCompareIon && expectedValue instanceof Decimal) ||
+        (!options.onlyCompareIon &&
+          (expectedValue instanceof ion.Decimal ||
+            expectedValue instanceof Number ||
+            typeof expectedValue === "number"))
+      )
+    ) {
+      return false;
+    }
```

Comment:
These code paths are (necessarily) pretty complicated. I think it would be nice to break these conditionals into separate cases that readers can digest independently and to which you can add separate comments.

```typescript
let isSupportedType = false;
let valueToCompare = null;
if (options.compareOnlyIon) {
  // `compareOnlyIon` requires that the provided value be an ion.dom.Decimal instance.
  if (expectedValue instanceof Decimal) {
    isSupportedType = true;
    valueToCompare = expectedValue.decimalValue();
  }
} else {
  // We will consider other Decimal-ish types
  if (expectedValue instanceof ion.Decimal) {
    // expectedValue is a non-DOM Decimal
    isSupportedType = true;
    valueToCompare = expectedValue;
  } else if (expectedValue instanceof Number || typeof expectedValue === "number") {
    isSupportedType = true;
    // calling numberValue() on ion.Decimal is lossy and could result in imprecise comparisons
    // hence converting number to ion.Decimal for comparison even though it maybe expensive
    valueToCompare = new ion.Decimal(expectedValue.toString());
  }
}

if (!isSupportedType) {
  return false;
}

return this.decimalValue().equals(valueToCompare);
```